src/Collections-Tests/WaitfreeQueueTest.class.st

```diff
@@ -0,0 +1,216 @@
+	"Answer a collection with the requested number of items"
+
+	| collection |
+
+	collection := OrderedCollection new.
+	count timesRepeat: [ collection add: AtomicQueueItem new ].
+	^collection
+]
+
+{ #category : #'helper methods' }
+WaitfreeQueueTest >> queueFromCollection: aCollection [
+	"Answer a WaitfreeQueue initialised from the supplied collection of items"
+
+	| queue |
+
+	queue := WaitfreeQueue new.
+	(queue instVarNamed: #dummy) next: aCollection first.
+	1 to: aCollection size - 1 do: [ :i |
+		(aCollection at: i) next: (aCollection at: i+1) ].
+	queue instVarNamed: #tail put: aCollection last.
+	^queue
+]
+
+{ #category : #tests }
+WaitfreeQueueTest >> testNextOrNil [
+
+	| queue |
+
+	queue := WaitfreeQueue new.
+	1 to: 3 do: [ :i |
+		queue 	nextPut: i ].
+
+	"Remove the entries"
+	1 to: 3 do: [ :i |
+		self assert: queue nextOrNil equals: i ].
+
+	"There should be 0 entries left with no zombies"
+	self assert: queue size equals: 0.
+	"But a zombie is always left behind"
+	self assert: queue itemCount equals: 1.
+
+	"Repeat to check that the number of zombies doesn't grow"
+	1 to: 3 do: [ :i |
+		queue 	nextPut: i ].
+
+	"Remove the entries"
+	1 to: 3 do: [ :i |
+		self assert: queue nextOrNil equals: i ].
+
+	"There should be 0 entries left with no zombies"
+	self assert: queue size equals: 0.
+	"But a zombie is always left behind"
+	self assert: queue itemCount equals: 1.
+
+]
+
+{ #category : #tests }
+WaitfreeQueueTest >> testNextOrNilSuchThat [
+
+	| queue |
+
+	queue := WaitfreeQueue new.
+	1 to: 10 do: [ :i |
+		queue 	nextPut: i ].
+
+	"Remove the first element"
+	self assert: (queue nextOrNilSuchThat: [ :o | o = 1 ]) equals: 1.
+
+	"Remove the last element"
+	self assert: (queue nextOrNilSuchThat: [ :o | o = 10 ]) equals: 10.
+
+	"Remove the middle elements"
+	4 to: 6 do: [ :i |
+		self assert: (queue nextOrNilSuchThat: [ :o | o = i ]) equals: i ].
+
+	"There should be 5 elements left with 1 zombie (the last element)"
+	self assert: queue size equals: 5.
+	"But... all items removed with #nextOrNilSuchThat: leave zombie behind"
+	self assert: queue itemCount equals: 6.
+	self assert: (queue instVarNamed: #tail) isZombie.
+
+
+	11 to: 13 do: [ :i |
+		queue 	nextPut: i ].
+
+	#(2 3 7 8 9 11 12 13) do: [ :i |
+		self assert: queue nextOrNil equals: i ].
+
+	"There should be 0 elements left with one zombie"
+	self assert: queue size equals: 0.
+	self assert: queue itemCount equals: 1.
+
+	queue nextPut: 11.
+	self assert: queue nextOrNil equals: 11.
+	self assert: queue nextOrNil equals: nil.
+
+	"There should be 0 elements left with one zombie"
+	self assert: queue size equals: 0.
+	self assert: queue itemCount equals: 1.
+
+]
+
+{ #category : #tests }
+WaitfreeQueueTest >> testZombieFirst [
+	"Check that a zombie item at the start of the queue is correctly removed"
+
+	| items queue real |
+
+	items := self newItems: 2.
+	queue := self queueFromCollection: items.
+	items first object: (queue instVarNamed: #dummy).
```

Comment:
It could be item method #beZombie or queue method #makeZombieAtItem:

Reply:
> It could be item method #beZombie or queue method #makeZombieAtItem:

An item can't make itself a zombie as it doesn't know what the dummy item is (the queue uses object identity to test).  It could be passed the dummy item, but that seems like extending the interface too far.

I've replaced the remaining uses of `#instVarNamed:`